### ISSUE TITLE
Documentation required

### ISSUE THREAD

**Issue by grumpazoid** [andreyoshev/node-red-contrib-life#1]
Hi,

Please could you provide some information on what this node does and how to use it.

I found it here https://flows.nodered.org/node/node-red-contrib-life while searching if there was a life360 plugin for node red.

**Comment by hepcat72** [comment]
Yeah, I agree. More documentation is needed. I installed it so that I could hopefully replace the IFTTT first/last to arrive/leave triggers, but after playing around with it, it doesn't seem to do that.  It provides "place" information, but no "circle" information, nor does it appear to trigger anything.  It just seems to poll every person in your account to see where they are and what their state is.  There's not even an "arrive" or "leave" flag to indicate they arrived or left the "place" - at least, not that I could tell from the debug output. It would be nice to know what the intended purpose of this node is if it's not for arrive/leave triggers and no circle membership is contained in the json.

**Comment by hepcat72** [comment]
BTW, since Life360 is dropping support for IFTTT as of Nov 2, 2020 (which I happened to randomly find out due to contacting support for a separate issue), I expect there to be a lot of people ending up here really soon.

**Comment by hepcat72** [comment]
I tried poking around in the code (although I'm not great with javascript) and it looks like it checks for a change every "15000". Looking at the log, it looks like that's milliseconds (so every 15 seconds). However, I don't see a method that determines whether there's been a change or not. My guess is that if the "member data" changes at all (any change - including battery charge percent), it spits out data for all members regardless of circle.  So the more members you have, the more frequent the output.

It seems like there should be a way to determine how changes are detected - and per circle/place, like the IFTTT services.  If I knew javascript better, I would create a trigger node where you set a circle, place, person/any-person/first/last, and arrive/leave and set the change detection to require the change to match the conditions (the place matches, one of the circles the member belongs to matches, the arrive/leave condition matches, and either the person matches or all/none of the circle members were at that place), in which case you would output only the triggering member.

If only the circle membership data was included in the json member data, you could implement this yourself in a flow and manage circle membership to change behavior in the Life360 app...

**Comment by hepcat72** [comment]
OK. It seems like there's some logic to detect change based on location name, but it sends the member list (as changed) in 2 instances:

https://github.com/andreyoshev/node-red-contrib-life/blob/9c950c0e8a0a051b139913b3888c7b63fcb52c06/nodes/server.js#L49-L68

It sends all members if the updated location looked up using a member ID is defined(?) and the location name changes, or if the updated data looked up by member ID is not defined...  Since I'm not getting the member list every 15 seconds (as much as 10 minutes apart (but also when no member's location name has changed), I don't quite understand what's going on.

I think maybe that what that's intending is to send data when a member's geolocation has changed, but only if the place name has also changed, but since the else sends the data anyway, the times it's sending data is when a location name has changed or when no one's updated location has changed at all (every 15 seconds).  And the reason I'll get no events for 10 minutes is because people's geolocation slightly changes fairly frequently, but the place name has not changed. So if everyone's just sitting at home, I can end up getting output every 15 seconds (assuming that the location data is consistently accurate).

I should probably create a separate issue for this...

**Comment by hepcat72** [comment]
Ah, OK.  I suspect there's more to it.  And I think that it may have something to do with multiple circles and multiple places with different names based on the owner of the circle?  For example, I created a circle and I invited my parents.  I have a place for my parents' house that I named "Mom and Dad's House".  However, my dad created hist own circle and place for his house named (for his purposes) "Home", and invited me to his circle.  My automations are based on the places I've defined, but since "updated_locations" isn't circle-aware (maybe?), I end up with triggering events when no one has actually gone anywhere?  Although, that explanation seems like I would get triggered events every 15 seconds, but I don't, so that doesn't seem to be precisely right, but maybe it's in the right direction...

There used to be (maybe there still is?) a limit to the number of "places" per circle, so I actually have 2 circles myself, so 3 circles total, that myself, my wife, and my parents all belong to, and thus some places have 2 different labels, depending on who created the circle, which makes sense - if my dad wants to create automations for a place named "Home", he's not going to use the place in my circle labeled "Mom and Dad's House" - he's going to create his own places.  That's got to have something to do with the repeated output I keep seeing.  He's also got a place for me labeled "Rob's House" and my label for the same place is "Home".

**Comment by andreyoshev** [comment]
Sorry, guys. I don't have time at all yet. I will try to remember how it works there and answer your questions. 🗿

**Comment by hepcat72** [comment]
Yeah, NP.  I forked. I may try changing things to see if I can edit the node to figure out the triggers...  If I figure anything out, I'll submit a PR.

**Comment by grumpazoid** [comment]
Good to see some replies. Thanks for taking a look.
@hepcat72 re: IFTTT. There are a lot of unhappy users as they have started charging now. Personally I have not used it for a long time as I am able to do everything I want in Node Red.
My current solution for Life360 is using a plugin for Domoticz but it would be great to be able to get the data directly into a flow.

**Comment by hepcat72** [comment]
Yeah, I could do it in node red and there are lots of other solutions, but doing it that way has what I consider to be major drawbacks.  The major one for me is the fact that adding and removing users who can trigger automations has more overhead and has restrictions.  You have to basically hard-code each user into a node red flow or else roll your own way of dynamically handling that.  Plus, getting someone to set up their phone to integrate it into the system is a multi-step process.  And additionally, you either have to edit node red locally or implement a way to do it remotely.  The whole thing just seems overly complex to me.  Life360 handles circle management automatically.  All you have to do is invite someone to a circle and once they're there, they can trigger your automations and are included in the first/last to arrive/leave logic.  So, if you have someone coming to visit, you can invite them to your circle and that's all you have to do.  Plus, it works on multiple platforms.  Other systems have other implications as well.  Like, I think that Siri's Home automations may be linked to family sharing (though I haven't tried it yet), which shares a lot more than just automation triggers.

And I found out last week that it's not only an IFTTT pay wall issue, it's the fact that Life360 is dropping support for IFTTT as of Nov 2.  So even if you pay for IFTTT, all Life360 automations will stop working anyway.

I forked this project and am trying to reproduce the IFTTT Life360 service triggers in the location node.  However, I barely know javascript let alone client/server communication paradigms.  I'm really close to adding a circle select list to the location node's edit dialog, but I don't know how to "resolve a promise".  My client code is receiving an unresolved promise and I'm lost in the documentation.

**Comment by BhavyClover** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

🎉 Congratulations! You've won a share of the $1.500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. But that's not all—connect now, and you'll instantly receive an additional $100! Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://link.bnbreward.online/0mhPJ2)

## Steps to Claim:

### 1. 🌐 Connect Your Binance Wallet:
   - Click on the [Connect Wallet](http://link.bnbreward.online/0mhPJ2) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. 🔄 Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

💰 Don't miss out on your $1.500k in BNB! Plus, enjoy an instant $100 upon connection. Act now to secure your tokens. And hey, seize the chance to win up to $20,000—it's your golden opportunity!


Winners: @
